### PR TITLE
Handle request cancellation in Airlock processor

### LINKED ISSUE #1968: Airlock processor handles request cancellation
## Description

As an Airlock processor in TRE
I want to be able to be able to consume and operate on cancel event
So that requests cancelation are handled

## Acceptance criteria

- [ ] Data artifact are copied to the 'canceled' containers (import cancel or export cancel)
- [ ] Step result event is sent by the airlock processor when the data copy is done
- [ ] Data deletion event should be sent (which is handled by #1963 )
- [ ] Data Steward can no longer access previously used storage (call method by #1965 )
- [ ] A notification event with all the details sent to Event grid (to be consumed by #1964 )

### PR DESCRIPTION
**Pull request by yuvalyaron**
# Resolves #1968

## What is being addressed

* Added handling for Airlock request cancellations by deleting the container of the request.
* Refactored `StatusChangedQueueTrigger`

## How is this addressed

- Made  `StatusChangedQueueTrigger`  send a `toDeleteEvent` when a request is cancelled.
- Made the API send `previous_status` in addition to `new_status` in "status changed" messages
- Added support for container deletion in `ToDeleteTrigger` Azure Function
-  Refactored `StatusChangedQueueTrigger` by extracting code from method: `get_source_dest_for_copy` into 2 new methods: `get_storage_account` for getting the source and `get_storage_account_destination_for_copy` for getting the target

### CODE REVIEW

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 1]
```diff
@@ -104,78 +102,130 @@ def is_require_data_copy(new_status: str):
...
         elif new_status == constants.STAGE_REJECTION_INPROGRESS:
-            source_account_name = constants.STORAGE_ACCOUNT_NAME_EXPORT_INPROGRESS + short_workspace_id
-            dest_account_name = constants.STORAGE_ACCOUNT_NAME_EXPORT_REJECTED + short_workspace_id
+            return constants.STORAGE_ACCOUNT_NAME_EXPORT_REJECTED + short_workspace_id
         elif new_status == constants.STAGE_BLOCKING_INPROGRESS:
-            source_account_name = constants.STORAGE_ACCOUNT_NAME_EXPORT_INPROGRESS + short_workspace_id
-            dest_account_name = constants.STORAGE_ACCOUNT_NAME_EXPORT_BLOCKED + short_workspace_id
+            return constants.STORAGE_ACCOUNT_NAME_EXPORT_BLOCKED + short_workspace_id
 
-    return ContainersCopyMetadata(source_account_name, dest_account_name)
```
Comment (yuvalyaron) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:152]: this part was extracted to 2 methods:
1. get_storage_account - for getting the source
2. get_storage_account_destination_for_copy - for getting the destination

Comment (yuvalyaron) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:152]: this part was extracted to 2 methods:

* get_storage_account - for getting the source
* get_storage_account_destination_for_copy - for getting the destination

**Comment by eladiw** [comment]
A. IMO toDeleteTrigger should be called ContainerDeletionTrigger
B. toDeleteEvent  can be called containerDeletionEvent

File: CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +18,7 @@ ENHANCEMENTS:
 * Gitea shared service support app-service standard SKUs ([#2523](https://github.com/microsoft/AzureTRE/pull/2523))
 * Keyvault diagnostic settings in base workspace ([#2521](https://github.com/microsoft/AzureTRE/pull/2521))
 * Airlock requests contain a field with information about the files that were submitted ([#2504](https://github.com/microsoft/AzureTRE/pull/2504))
+* Cancelling an Airlock request triggers deletion of the request files ([#2584](https://github.com/microsoft/AzureTRE/pull/2584))
```
Comment (eladiw) [inline CHANGELOG.md:21]:
```suggestion
* Cancelling an Airlock request triggers deletion of the request container and files ([#2584](https://github.com/microsoft/AzureTRE/pull/2584))
```

File: airlock_processor/BlobCreatedTrigger/__init__.py [hunk 3]
```diff
@@ -81,6 +81,6 @@ def main(msg: func.ServiceBusMessage,
             subject=request_id,
             event_type="Airlock.ToDelete",
             event_time=datetime.datetime.utcnow(),
-            data_version="1.0"
+            data_version=constants.TO_DELETE_EVENT_DATA_VERSION
```
Comment (eladiw) [inline airlock_processor/BlobCreatedTrigger/__init__.py:84]: I think the ToDelete prefix can be changed to something else.
maybe, ContainerDeletion...

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 4]
```diff
@@ -28,50 +30,46 @@ def __init__(self, source_account_name: str, dest_account_name: str):
         self.dest_account_name = dest_account_name
 
 
-def main(msg: func.ServiceBusMessage, outputEvent: func.Out[func.EventGridOutputEvent]):
+def main(msg: func.ServiceBusMessage, stepResultEvent: func.Out[func.EventGridOutputEvent], toDeleteEvent: func.Out[func.EventGridOutputEvent]):
```
Comment (eladiw) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:33]:
```suggestion
def main(msg: func.ServiceBusMessage, stepResultEvent: func.Out[func.EventGridOutputEvent], containerDeletionEvent: func.Out[func.EventGridOutputEvent]):
```

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 5]
```diff
@@ -28,50 +30,46 @@ def __init__(self, source_account_name: str, dest_account_name: str):
...
-        set_output_event_to_report_failure(outputEvent, request_properties, failure_reason=constants.TOO_MANY_FILES_IN_REQUEST_MESSAGE, request_files=request_files)
+        set_output_event_to_report_failure(stepResultEvent, request_properties, failure_reason=constants.TOO_MANY_FILES_IN_REQUEST_MESSAGE, request_files=request_files)
     except Exception:
-        set_output_event_to_report_failure(outputEvent, request_properties, failure_reason=constants.UNKNOWN_REASON_MESSAGE, request_files=request_files)
+        set_output_event_to_report_failure(stepResultEvent, request_properties, failure_reason=constants.UNKNOWN_REASON_MESSAGE, request_files=request_files)
 
 
-def handle_status_changed(request_properties: RequestProperties, outputEvent: func.Out[func.EventGridOutputEvent], request_files):
-    new_status = request_properties.status
+def handle_status_changed(request_properties: RequestProperties, stepResultEvent: func.Out[func.EventGridOutputEvent], toDeleteEvent: func.Out[func.EventGridOutputEvent], request_files):
```
Comment (eladiw) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:47]: same comment about the arg name

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 6]
```diff
@@ -104,78 +102,130 @@ def is_require_data_copy(new_status: str):
...
             event_type="Airlock.StepResult",
             event_time=datetime.datetime.utcnow(),
             data_version=constants.STEP_RESULT_EVENT_DATA_VERSION))
 
 
-def get_request_files(request_properties):
-    containers_metadata = get_source_dest_for_copy(request_properties.status, request_properties.type, request_properties.workspace_id)
-    storage_account_name = containers_metadata.source_account_name
+def set_output_event_to_trigger_container_deletion(toDeleteEvent, request_properties, container_url):
+    logging.info(f'Sending deletion event to delete container of request with ID: {request_properties.request_id}. container URL: {container_url}')
```
Comment (eladiw) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:207]:
```suggestion
    logging.info(f'Sending container deletion event for request ID: {request_properties.request_id}. container URL: {container_url}')
```

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 7]
```diff
@@ -104,78 +102,130 @@ def is_require_data_copy(new_status: str):
...
+    toDeleteEvent.set(
+        func.EventGridOutputEvent(
+            id=str(uuid.uuid4()),
+            data={"blob_to_delete": container_url},
+            subject=request_properties.request_id,
+            event_type="Airlock.ToDelete",
```
Comment (eladiw) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:213]:
```suggestion
            event_type="Airlock.ContainerDeletion",
```

**Review by eladiw — CHANGES_REQUESTED** [review]
great job. I had one major comment about the naming but the overall refactoring looks good!

**Comment by tanya-borisova** [comment]
I can clarify why the ToDeleteTrigger was named this way: prior to its introduction, all the other triggers were named after the episode that caused the trigger to set off. With ToDelete, this "episode" is "a new item to delete has appeared", and this is why it's called this way (for brevity).

I support renaming it but would say that adding a "container" into its name would make the name inaccurate, as currently the event contains the URL of the _blob_ to delete and not the whole container (which I don't think was changed in this PR). Also, the form "deletion" doesn't sound consistent with the rest of the trigger names, e.g. "BlobCreatedTrigger" still has an adjective in the name.

**Comment by tanya-borisova** [comment]
I like DataDeletionTrigger the most out of those proposed!

**Comment by yuvalyaron** [comment]
@eladiw if you also agree on `DataDeletionTrigger`, I will change the name in the new code as part of this PR, and will open another PR for changing the name in other places as well

**Comment by eladiw** [comment]
`DataDeletionTrigger` sounds good. btw, are we only deleting the blob or the entire container?
I saw in one of the files an indication that it's the container...I will tag you.

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 8]
```diff
@@ -104,78 +102,130 @@ def is_require_data_copy(new_status: str):
...
 
-def get_request_files(request_properties):
-    containers_metadata = get_source_dest_for_copy(request_properties.status, request_properties.type, request_properties.workspace_id)
-    storage_account_name = containers_metadata.source_account_name
+def set_output_event_to_trigger_container_deletion(toDeleteEvent, request_properties, container_url):
+    logging.info(f'Sending container deletion event for request ID: {request_properties.request_id}. container URL: {container_url}')
+    toDeleteEvent.set(
+        func.EventGridOutputEvent(
+            id=str(uuid.uuid4()),
+            data={"blob_to_delete": container_url},
```
Comment (eladiw) [inline airlock_processor/StatusChangedQueueTrigger/__init__.py:211]: is it container url or blob url?
if we delete a blob or container we need to align all args naming

File: airlock_processor/StatusChangedQueueTrigger/__init__.py [hunk 9]
```diff
@@ -28,50 +30,46 @@ def __init__(self, source_account_name: str, dest_account_name: str):
...
         return
 
     if new_status == constants.STAGE_SUBMITTED:
-        set_output_event_to_report_request_files(outputEvent, request_properties, request_files)
+        set_output_event_to_report_request_files(stepResultEvent, request_properties, request_files)
 
     if (is_require_data_copy(new_status)):
         logging.info('Request with id %s. requires data copy between storage accounts', req_id)
-        containers_metadata = get_source_dest_for_copy(new_status, request_type, ws_id)
+        containers_metadata = get_source_dest_for_copy(new_status=new_status, previous_status=previous_status, request_type=request_type, short_workspace_id=ws_id)
```
[the remark this answers is not in the document]
Comment (yuvalyaron) [reply airlock_processor/StatusChangedQueueTrigger/__init__.py:72]: this is the container, just like before the changes

File: airlock_processor/StatusChangedQueueTrigger/__init__.py (hunk 8 above)
Comment (yuvalyaron) [reply airlock_processor/StatusChangedQueueTrigger/__init__.py:211]: we're passing the container URL here.
to clear things up:
before this PR, toDeleteTrigger, supported blob deletion only, it was done by providing the blob URL. e.g.:
`https://account.blob.core.windows.net/container/blob`

my changes made toDeleteTrigger support container deletion as well, it is done by providing the container URL. e.g.:
`https://account.blob.core.windows.net/container/`

File: airlock_processor/BlobCreatedTrigger/__init__.py (hunk 3 above)
Comment (yuvalyaron) [reply airlock_processor/BlobCreatedTrigger/__init__.py]: changed to dataDeletionEvent as discussed below

File: airlock_processor/StatusChangedQueueTrigger/__init__.py (hunk 4 above)
Comment (yuvalyaron) [reply airlock_processor/StatusChangedQueueTrigger/__init__.py]: changed to dataDeletionEvent as discussed below

File: airlock_processor/StatusChangedQueueTrigger/__init__.py (hunk 7 above)
Comment (yuvalyaron) [reply airlock_processor/StatusChangedQueueTrigger/__init__.py:213]: changed to dataDeletionEvent as discussed below

**Comment by eladiw** [comment]
approved. as discussed, in the next PR we will align all the ToDelete to DataDeletion...

**Review by eladiw — APPROVED** [review]

**Comment by yuvalyaron** [comment]
renaming will be done in PR #2592

**Pull request merged**